File: python/TheoryOfComputation/BackusNaurForm/Tree.py
from __future__ import annotations
from typing import Generic, TypeVar, List

T = TypeVar('T')


class Tree(Generic[T]):
    __value: T
    __children: List[Tree[T]]

    def __init__(self, value: T):
        self.__value = value
        self.__children = []

    def __repr__(self):
        if len(self.__children) > 0:
            return "{} -> [{}]".format(self.__value, ",".join(str(x) for x in self.__children))
        else:
            return str(self.__value)

    def add_child_value(self, value: T) -> Tree[T]:
        return self.add_child(Tree(value))

    def add_child(self, child: Tree[T]) -> Tree[T]:
        self.__children.append(child)
        return self

    def get_value(self):
        return self.__value

    def get_children(self) -> List[Tree[T]]:
        return self.__children

File: lib/src/dataStructures/tree/Tree.ts
import { Consumer, ToString } from "../../types";

class Tree<T> {
    value: T;
    children: Tree<T>[];

    constructor(value: T) {
        this.value = value;
        this.children = [];
    }

    visit(visitor: Consumer<T>) {
        visitor(this.value);
        this.children.forEach(child => child.visit(visitor));
    }

    toString(valueToString: ToString<T>): string {
        if (this.children.length > 0) {
            return `${valueToString(this.value)} -> [${this.children.map(c => c.toString(valueToString)).join(',')}]`
        } else {
            return valueToString(this.value);
        }
    }

    addChildValue(value: T): this {
        return this.addChild(new Tree(value));
    }

    addChild(child: Tree<T>): this {
        this.children.push(child);
        return this;
    }

    getValue() {
        return this.value;
    }

    getChildren(): Tree<T>[] {
        return this.children;
    }
}

export default Tree;